README.md
# RISC-V Compliance Framework
The work here is under development and is not stable for consumption .. yet!

## Documentation
You can find the latest documentation of the work: [here](https://riscof.readthedocs.io/)

Sample command for riscof

Spike vs Spike
```
python -m riscof.main -bm model_from_yaml -bf Examples/template_env.yaml -eyaml Examples/template_env.yaml -dm model_from_yaml -ispec Examples/rv32i_isa.yaml -pspec Examples/rv32i_platform.yaml --verbose debug


```

Eclass vs Spike
```
python -m riscof.main -bm model_from_yaml -bf Examples/template_env.yaml -dm model_from_yaml -ispec Examples/rv32i_isa.yaml -pspec Examples/rv32i_platform.yaml -eyaml Examples/e_class_env.yaml --verbose debug

```

dbgen
```
python -m riscof.dbgen

```


CHANGELOG.md
# CHANGELOG

This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

Please note the header `WIP-DEV` is to always remain indicating the changes done on the dev branch.
Only when a release to the main branch is done, the contents of the WIP-DEV are put under a
versioned header while the `WIP-DEV` is left empty

## [WIP-DEV]
- Ensure correct riscof return code when `--no-browser` is used. Fixes #87.
- Updating CONTRIBUTING.rst to capture the new git strategy adopted to follow a monthly release
- Fix typo in installation.rst
- Remove unused pip import in setup.py

## [1.25.3] - 2023-01-24
- use "make -k" in riscof_model.py template to ensure all test cases run, even after a failure. Fixes #73.

## [1.25.2] - 2022-10-08
- updating paths for docker image and plugins to point to new location

## [1.25.1] - 2022-09-09
- Modified the code to fix the issue #62, which changed the condition to allow flen=64 for D extension tests.

## [1.25.0] - 2022-09-07
- migrated to using riscv-config version 3.2.0+
- modified functions to use the new warl_class from riscv-config-3.2.0+
- modified prod_isa to use riscv-config isa_validator

## [1.24.3] - 2022-08-29
- fix typeo in doc for command related to cloning arch-suites
- bump riscv-config version to be 2.17.0

## [1.24.2] - 2022-07-27
- Fix `riscv_config` dependency version to 2.10.1

## [1.24.1] - 2022-07-19
- Account for the same test to be included with both XLEN variants in the isa generation.
- Add markdown report for coverage statistics.

## [1.24.0] - 2022-05-14
- rename the "master" branch of riscv-arch-test to "main"  

## [1.23.4] - 2022-02-24
- fixed doc dependency on mistune for successful RTD build

## [1.23.3] - 2022-01-27
- Added missing subextensions for B and K specs in ISA generation function.

## [1.23.2] - 2021-11-23
- Fixed report generation for failing tests. Removed HTML table and added manual diff of lines in the report.
- Added no_clean flag to CLI. 

## [1.23.1] - 2021-10-20
- Improved diff reporting in case of Failed tests.
- Added support for f and d exts in model plugin.
- Added checking to ensure that tests with wrong filtering based on ISA are not selected.
- Fixed spelling for opening and arch test command in docs.

## [1.23.0] - 2021-10-14
- Added support for new RVTEST_ISA macro
- Fixed decode error in Command util.

## [1.22.1] - 2021-09-13
- Added return code of 1 on error

## [1.22.0] - 2021-09-10
- Fixed signature alignment to begin and end at 16-byte boundaries for all header templates.

## [1.21.2] - 2021-09-03
- Ported CLI to use click package
- Improved documentation for using plugins.
- Added separate paths for dut and ref artifacts.

## [1.21.1] - 2021-07-24
- Added support for intermediate run start points #1
- --dbfile is added to generate the test-list and then run tests on the target and reference
- --testfile is added to directly run tests on the target and reference (assumes the testlist has been correctly generated)
- --no-ref-run is added so that tests do not run on reference and quit before signature comparison
- --no-dut-run is added so that tests do not run on DUT and quit before signature comparison 
- Added a timestamp comment on the generated database.yaml and testlist.yaml files
- Change FXLEN macro to FLEN #8

## [1.21.0] - 2021-07-21
- Changed CI script from gitlab to github actions
- Removing hosted cgf files
- Removing hosted riscv-test-suite
- Removing hosted database.yaml
- Updated report to capture the version of the riscv-arch-test suite used during run
- Added new cli - 'arch-test' to automatically clone, update and maintain the riscv-arch-test repo
- Using annotated tags for CI releases
- `suite` and `env` are now required cli args
- updated docs 
  - added new sections on commands and inputs. 
  - Revised the installation and riscv-arch-test sections. 
  - Fixed testlist and dbgen usage and formats
  - Cleaned up examples. 
  - Removed testformat spec.
  - Removed section on adding new tests.
  - Adding new doc on available PLUGINS for RISCOF.
  - Fixed links for new repo transition


## [1.20.3] - 2021-06-30
- Updated docs with guide on running riscv-arch-tests with RISCOF

## [1.20.2] - 2021-06-30
- Fixed ISA check typo for 64-bit K extension tests and database.yaml
- Fixed the issue pointed out in #53. Now riscof will through an error if test_list.yaml is empty and will not proceed furthur.

## [1.20.1] - 2021-06-24
- Added env argument to the cli.
- printing versions of isac and confing in log for better debugging
- fixed issues pointed out in #51
- fixed regex conditions for cebreak tests and also updated database
- Updated tests based on latest coverpoints.

## [1.20.0] - 2021-05-30
- work directory is now controlled by user args. Defaults to riscof_work for backward compatibility.
- when filtering we check the config string in the cgf. The config node can have multiple entries. 
  If within an entry multiple conditions are detected, then they are andded to create included_entry. 
  Across all the list of entries in the config list the included_entries are ORed.

## [1.19.0] - 2021-04-24
- Added new rv64i and rv32i tests from arch-test repo
- Added K_unratified tests from arch-test repo
- Addded all relevant coverage files 
- Updated coverage support to pick local coverage files
- Updated setup phase to dump out sail_cSim reference model
- Updated the config.ini and model setup files
- Updated docs

## [1.18.4] - 2021-01-27
- Fixed missing coverage.html during pip installation.

## [1.18.3] - 2021-01-21
- Added support for paths relative to `config.ini` location
- Plugins get the location of `config.ini` directory with the key `config_dir`

## [1.18.2] - 2021-01-14
- Fixed missing files in PYPI package
- Added `physical_addr_sz` field in model_isa.yaml

## [1.18.1] - 2021-01-11
- updated ci to use latest twine
- using api-tokens for twine uploads

## [1.18.0] - 2021-01-11
- updated cli to accept multiple cgf files similar to isac

suite/MUL.S
# RISC-V Compliance Test RV32IM-MUL-01
#
# Copyright (c) 2018, Imperas Software Ltd.
# All rights reserved.
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions are met:
#      * Redistributions of source code must retain the above copyright
#        notice, this list of conditions and the following disclaimer.
#      * Redistributions in binary form must reproduce the above copyright
#        notice, this list of conditions and the following disclaimer in the
#        documentation and/or other materials provided with the distribution.
#      * Neither the name of the Imperas Software Ltd. nor the
#        names of its contributors may be used to endorse or promote products
#        derived from this software without specific prior written permission.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
# IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
# THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
# PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL Imperas Software Ltd. BE LIABLE FOR ANY
# DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
# (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
# ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
# THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
#
# Specification: RV32IM Base Integer Instruction Set, Version 2.0
# Description: Testing instruction MUL.

#include "test_macros.h"
#include "compliance_test.h"
#include "compliance_io.h"

RV_COMPLIANCE_RV32M

RV_COMPLIANCE_CODE_BEGIN

	RVTEST_IO_INIT
	RVTEST_IO_ASSERT_GPR_EQ(x0, 0x00000000)
	RVTEST_IO_WRITE_STR("Test Begin Reserved regs ra(x1) a0(x10) t0(x5)\n")
  RVTEST_START

	# ---------------------------------------------------------------------------------------------

	RVTEST_IO_WRITE_STR("# Test number 1 - corner cases\n")

  #ifdef TEST_PART_1
    RVTEST_PART_START(1, "# Corner Case -1")
	    TEST_RR_SRC2(mul, x17, x18, 0, 0x0, 0x0, x2, 0)
    RVTEST_PART_END(1)
  #endif
  #ifdef TEST_PART_2
    RVTEST_PART_START(2, "# Corner Case -1")
	    TEST_RR_SRC2(mul, x19, x20, 0, 0x0, 0x1, x2, 4)
    RVTEST_PART_END(2)
  #endif
  #ifdef TEST_PART_3
    RVTEST_PART_START(3, "# Corner Case -1")
	    TEST_RR_SRC2(mul, x21, x22, 0, 0x0, -0x1, x2, 8)
    RVTEST_PART_END(3)
  #endif
  #ifdef TEST_PART_4
    RVTEST_PART_START(4, "# Corner Case -1")
	    TEST_RR_SRC2(mul, x23, x24, 0, 0x0, 0x7fffffff, x2, 12)
    RVTEST_PART_END(4)
  #endif
  #ifdef TEST_PART_5
    RVTEST_PART_START(5, "# Corner Case -1")
	    TEST_RR_SRC2(mul, x25, x26, 0, 0x0, 0x80000000, x2, 16)
    RVTEST_PART_END(5)
  #endif

	# ---------------------------------------------------------------------------------------------

	RVTEST_IO_WRITE_STR("# Test number 2 - corner cases\n")

  #ifdef TEST_PART_6
    RVTEST_PART_START(6, "# Corner Case -1")
	    TEST_RR_SRC2(mul, x27, x28, 0, 0x1, 0x0, x2, 0)
    RVTEST_PART_END(6)
  #endif
  #ifdef TEST_PART_7
    RVTEST_PART_START(7, "# Corner Case -1")
	    TEST_RR_SRC2(mul, x29, x30, 0x1, 0x1, 0x1, x2, 4)
    RVTEST_PART_END(7)
  #endif
  #ifdef TEST_PART_8
    RVTEST_PART_START(8, "# Corner Case -1")
	    TEST_RR_SRC2(mul, x31, x3, 0xffffffff, 0x1, -0x1, x2, 8)
    RVTEST_PART_END(8)
  #endif
  #ifdef TEST_PART_9
    RVTEST_PART_START(9, "# Corner Case -1")
	    TEST_RR_SRC2(mul, x4, x8, 0x7fffffff, 0x1, 0x7fffffff, x2, 12)
    RVTEST_PART_END(9)
  #endif
  #ifdef TEST_PART_10
    RVTEST_PART_START(10, "# Corner Case 0")
	    TEST_RR_SRC2(mul, x9, x11, 0x80000000, 0x1, 0x80000000, x2, 16)
    RVTEST_PART_END(10)
  #endif

	# ---------------------------------------------------------------------------------------------

	RVTEST_IO_WRITE_STR("# Test number 3 - corner cases\n")

  #ifdef TEST_PART_11
    RVTEST_PART_START(11, "# Corner Case 0")
	    TEST_RR_SRC2(mul, x12, x13, 0, -0x1, 0x0, x2, 0)
    RVTEST_PART_END(11)
  #endif
  #ifdef TEST_PART_12
    RVTEST_PART_START(12, "# Corner Case 0")
	    TEST_RR_SRC2(mul, x14, x15, 0xffffffff, -0x1, 0x1, x2, 4)
    RVTEST_PART_END(12)
  #endif
  #ifdef TEST_PART_13
    RVTEST_PART_START(13, "# Corner Case 0")
	    TEST_RR_SRC2(mul, x16, x17, 0x1, -0x1, -0x1, x2, 8)
    RVTEST_PART_END(13)
  #endif
  #ifdef TEST_PART_14
    RVTEST_PART_START(14, "# Corner Case 0")
	    TEST_RR_SRC2(mul, x18, x19, 0x80000001, -0x1, 0x7fffffff, x2, 12)
    RVTEST_PART_END(14)
  #endif
  #ifdef TEST_PART_15
    RVTEST_PART_START(15, "# Corner Case 0")
	    TEST_RR_SRC2(mul, x20, x21, 0x80000000, -0x1, 0x80000000, x2, 16)
    RVTEST_PART_END(15)
  #endif

	# ---------------------------------------------------------------------------------------------

	RVTEST_IO_WRITE_STR("# Test number 4 - corner cases\n")

  #ifdef TEST_PART_16
    RVTEST_PART_START(16, "# Corner Case 0")
	    TEST_RR_SRC2(mul, x22, x23, 0, 0x7fffffff, 0x0, x2, 0)
    RVTEST_PART_END(16)
  #endif
  #ifdef TEST_PART_17
    RVTEST_PART_START(17, "# Corner Case 0")
	    TEST_RR_SRC2(mul, x24, x25, 0x7fffffff, 0x7fffffff, 0x1, x2, 4)
    RVTEST_PART_END(17)
  #endif
  #ifdef TEST_PART_18
    RVTEST_PART_START(18, "# Corner Case 0")
	    TEST_RR_SRC2(mul, x26, x27, 0x80000001, 0x7fffffff, -0x1, x2, 8)
    RVTEST_PART_END(18)
  #endif
  #ifdef TEST_PART_19
    RVTEST_PART_START(19, "# Corner Case 0")
	    TEST_RR_SRC2(mul, x28, x29, 0x1, 0x7fffffff, 0x7fffffff, x2, 12)
    RVTEST_PART_END(19)
  #endif
  #ifdef TEST_PART_20
    RVTEST_PART_START(20, "# Corner Case 0")
	    TEST_RR_SRC2(mul, x30, x31, 0x80000000, 0x7fffffff, 0x80000000, x2, 16)
    RVTEST_PART_END(20)
  #endif

	# ---------------------------------------------------------------------------------------------

	RVTEST_IO_WRITE_STR("# Test number 5 - corner cases\n")

  #ifdef TEST_PART_21
    RVTEST_PART_START(21, "# Corner Case 0")
	    TEST_RR_SRC2(mul, x3, x4, 0, 0x80000000, 0x0, x2, 0)
    RVTEST_PART_END(21)
  #endif
  #ifdef TEST_PART_22
    RVTEST_PART_START(22, "# Corner Case 0")
	    TEST_RR_SRC2(mul, x8, x9, 0x80000000, 0x80000000, 0x1, x2, 4)
    RVTEST_PART_END(22)
  #endif
  #ifdef TEST_PART_23
    RVTEST_PART_START(23, "# Corner Case 0")
	    TEST_RR_SRC2(mul, x11, x12, 0x80000000, 0x80000000, -0x1, x2, 8)
    RVTEST_PART_END(23)
  #endif
  #ifdef TEST_PART_24
    RVTEST_PART_START(24, "# Corner Case 0")
	    TEST_RR_SRC2(mul, x13, x14, 0x80000000, 0x80000000, 0x7fffffff, x2, 12)
    RVTEST_PART_END(24)
  #endif
  #ifdef TEST_PART_25
    RVTEST_PART_START(25, "# Corner Case 0")
	    TEST_RR_SRC2(mul, x15, x16, 0, 0x80000000, 0x80000000, x2, 16)
    RVTEST_PART_END(25)
  #endif

	RVTEST_IO_WRITE_STR("Test End\n")

	# ---------------------------------------------------------------------------------------------
test_end:
	RV_COMPLIANCE_HALT

RV_COMPLIANCE_CODE_END

# Input data section.
	.data

# Output data section.
RV_COMPLIANCE_DATA_BEGIN
test_res:
	.fill 4, 4, -1
	.fill 4, 4, -1
	.fill 4, 4, -1
	.fill 4, 4, -1
	.fill 4, 4, -1
	.fill 4, 4, -1
	.fill 1, 4, -1

RV_COMPLIANCE_DATA_END
